core/src/main/java/org/apache/iceberg/hadoop/JdbcTableDao.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class JdbcTableDao {
+  public static final String TABLES_TABLE_NAME = "iceberg_tables";
+  public static final String JDBC_CATALOG_TABLE_DDL =
+          "CREATE TABLE IF NOT EXISTS " + JdbcTableDao.TABLES_TABLE_NAME +
+                  "(catalogName VARCHAR(1255) NOT NULL," +
+                  "tableNamespace VARCHAR(1255) NOT NULL," +
+                  "tableName VARCHAR(1255) NOT NULL," +
+                  "metadataLocation VARCHAR(32768)," +
+                  "previousMetadataLocation VARCHAR(32768)," +
+                  "PRIMARY KEY (catalogName,tableNamespace,tableName)," +
+                  "FOREIGN KEY (catalogName, tableNamespace) REFERENCES " + JdbcNamespaceDao.NAMESPACES_TABLE_NAME +
+                  "(catalogName, namespace)" +
+                  ");";
+  private static final Logger LOG = LoggerFactory.getLogger(JdbcTableDao.class);
+  private final QueryRunner queryRunner;
+  private final String catalogName;
+
+
+  public JdbcTableDao(QueryRunner queryRunner, String catalogName) {
+    this.queryRunner = queryRunner;
+    this.catalogName = catalogName;
+  }
+
+  public boolean isExists(TableIdentifier tableIdentifier) {
+    try {
+      return this.get(tableIdentifier) != null;
+    } catch (SQLException throwables) {
+      return false;
+    }
+  }
+
+  public JdbcTable get(TableIdentifier tableIdentifier) throws SQLException {
+    return queryRunner.query("SELECT * FROM " + TABLES_TABLE_NAME + " " +
+                    "WHERE catalogName = ? AND tableNamespace = ? AND tableName = ? ",
+            new BeanHandler<>(JdbcTable.class),
+            catalogName, tableIdentifier.namespace().toString(), tableIdentifier.name());
+  }
+
+  public List<TableIdentifier> getAll(Namespace namespace) throws SQLException {
+    List<JdbcTable> tables = queryRunner.query("SELECT * FROM " + TABLES_TABLE_NAME + " " +
+                    "WHERE catalogName = ? AND tableNamespace = ?",
+            new BeanListHandler<>(JdbcTable.class), catalogName, namespace.toString());
+    List<TableIdentifier> results = Lists.newArrayList();
+    for (JdbcTable table : tables) {
+      results.add(table.toTableIdentifier());
+    }
+    return results;
+  }
+
+  public void update(TableIdentifier tableIdentifier, String oldMetadataLocation,
+                     String newMetadataLocation) throws SQLException {
+    queryRunner.update("UPDATE " + TABLES_TABLE_NAME + " SET metadataLocation = ? , previousMetadataLocation = ? " +
+                    "WHERE catalogName = ? AND tableNamespace = ? " +
+                    "AND tableName = ? ", newMetadataLocation,
+            oldMetadataLocation, catalogName, tableIdentifier.namespace().toString(), tableIdentifier.name());
```

Comment:
Looks like this doesn't yet implement an atomic swap of the old metadata location for the new one, so table updates would be unsafe because concurrent writers would clobber each other's commits. @kbendick, did your solution tackle that problem yet?

Reply:
I think all you would need to do is to change the SQL slightly:

```sql
UPDATE iceberg_tables
SET metadata_location = ?, previous_metadata_location = ?
WHERE table_namespace = ? AND table_name = ? AND metadata_location = ?
```

That adds a predicate to ensure that the table namespace and name combination will only be changed if the metadata_location is the expected one. If that affects 0 rows, then the metadata location was changed concurrently. If it affects 1 row, the commit was successful, and if it affects more than one row, there's a problem. (Note that I've simplified this to omit `catalog_name`.)

Reply:
changed the update statement if update statement effects >0 rows it will succeed otherwise will throw `CommitFailedException` exception.

Reply:
updated the logic, now it will 
for committing table: 
- succeed only if it effects row =1 
- else rollback with `CommitFailedException`

for renaming table: 
- succeed only if it effects row = 1 
- rollback with `NoSuchTableException`  if it effects row =0
- else rollback fail with `UncheckedIOException`

Reply:
to strengthen it further we can user[ row level locking](https://stackoverflow.com/a/46995306/2679740) 'SELECT FOR UPDATE'
if its not implemented JDBC Driver Throws `SQLFeatureNotSupportedException`. using this exception we could fallback to regular update statement.